### PR TITLE
[docs] Update configuration-ini.md file

### PR DESCRIPTION
**Pull request by songgot**
- Remove learning_rate from oprimizer section
- Add learning rate scheduler section
- Remove dataset section
- Add train set, validation set and test set section
- Add many types to layer section
- Add table about type, key, value, default value and description for each layers
- Update configuration file example

**Self evaluation:**
1. Build test:   [X]Passed [ ]Failed [ ]Skipped
2. Run test:     [X]Passed [ ]Failed [ ]Skipped

Signed-off-by: Hyunil <hyunil46.park@samsung.com>

### CODE REVIEW

File: docs/configuration-ini.md [hunk 1]
```diff
@@ -13,9 +13,9 @@ If you want more about iniparser, please visit
 
 Configuration file consists of Two Sections, Network and Layer
```
Comment (jijoongmoon) [inline docs/configuration-ini.md:14]: How about delete this line? There are more than Network and Layer.

File: docs/configuration-ini.md [hunk 2]
```diff
@@ -101,159 +98,395 @@ beta2 = 0.999
...
-
-2. ```traindata = <string>```
-
-    training data file path.   The data must be saved as following
-
-    ```feature data[i], label data[i], feature data[i+1], label data[i+1], ...```
+### Learning Rate Scheduler Section
 
-3. ```validdata = <string>```
+Define the learing rate, decay steps and decay rate.
```
Comment (jijoongmoon) [inline docs/configuration-ini.md:104]: How about add 
```
Start with "[LearningRateScheduler]"
```

File: docs/configuration-ini.md [hunk 3]
```diff
@@ -101,159 +98,395 @@ beta2 = 0.999
...
+type=constant
+Learning_rate = 1e-4 	# Learning Rate
+```
 
-   shape of input (shouldn't be zero).
+### Train Set Section
 
-7. ```activation = <string>```
+Define the type and path of the traing data file.
+Start with "[train_set]"
```
Comment (jijoongmoon) [inline docs/configuration-ini.md:129]: add new line here.

File: docs/configuration-ini.md [hunk 4]
```diff
@@ -101,159 +98,395 @@ beta2 = 0.999
...
+   Currently only file is supported.
 
-     set filters size for convolution layer
+2. ``` path = <string> ```
 
-13. ```stride = <unsigned int>,<unsigned int>```
+   Data path for validation.
 
-     set stride for convolution and pooling layer
+Below is a sample TrainSet section.
```
Comment (jijoongmoon) [inline docs/configuration-ini.md:161]: %s/TrainSet/ValidationSet/

File: docs/configuration-ini.md [hunk 5]
```diff
@@ -101,159 +98,395 @@ beta2 = 0.999
...
 
-14. ```padding = <unsigned int>,<unsigned int>```
+```ini
+# TrainSet Section
+[train_set]
```
Comment (jijoongmoon) [inline docs/configuration-ini.md:165]: %s/train/valid

File: docs/configuration-ini.md [hunk 6]
```diff
@@ -101,159 +98,395 @@ beta2 = 0.999
...
+Type = file
+path = valDataset.dat
+```
 
-     set padding for convolution and pooling layer
+### Test Set Section
 
-15. ```pool_size = <unsigned int>,<unsigned int>```
+Define the type and path of the test data file.
+Start with "[valid_set]"
```
Comment (jijoongmoon) [inline docs/configuration-ini.md:173]: %s/valid/test/

File: docs/configuration-ini.md [hunk 7]
```diff
@@ -101,159 +98,395 @@ beta2 = 0.999
...
-       * average : average pooling
-       * global_max : global max pooling
-       * global_average : global average pooling
+2. ``` path = <string> ```
 
-17. ```flatten = <bool>```
+   Data path for test.
 
-    flattens the output of this layer.
+Below is a sample TrainSet section.
```
Comment (jijoongmoon) [inline docs/configuration-ini.md:183]: %s/TrainSet/TestSet/

File: docs/configuration-ini.md [hunk 8]
```diff
@@ -101,159 +98,395 @@ beta2 = 0.999
...
-
-2. ```kernel_size = <unsigned int>,<unsigned int>```
-
-   Kernel size for convolution layer
-
-3. ```bias_init_zero = <bool>```
+   Initial learning rate to decay
 
-   token to initialize bias with zeros. Setting to False would initialize bias randomly.
+2. ```decay_steps = <float> ```
```
Comment (lhs8928) [inline docs/configuration-ini.md:109]: Please describe that decay_steps, decay_rate is used in exponential scheduler.
And also can you please describe properties for step scheduler?

File: docs/configuration-ini.md [hunk 9]
```diff
@@ -101,159 +98,395 @@ beta2 = 0.999
...
+     * concat : concat layer
+     * multiout : multiout layer
+     * embedding : embedding layer
+     * rnn : RNN layer
+     * lstm : LSTM layer
+     * split : split layer
+     * gru : GRU layer
+     * permute : permute layer
+     * dropout : dropout layer
+     * backbone_nnstreamer : backbone layer using nnstreamer
```
Comment (jijoongmoon) [inline docs/configuration-ini.md:217]: NNTrainer also suppors backbone_tflite as well. HOw about adding it?

File: docs/configuration-ini.md [hunk 10]
```diff
@@ -344,71 +577,91 @@ activation = softmax
...
+
+[Optimizer]
+Type = adam
+beta1 = 0.9       # beta 1 for adam
+beta2 = 0.999     # beta 2 for adam
+epsilon = 1e-7    # epsilon for adam
+
+[LearningRateScheduler]
+type=constant
+Learning_rate = 1e-4 # Learning Rate
```
Comment (jijoongmoon) [inline docs/configuration-ini.md:604]: How about adding train_set for data set?

File: docs/configuration-ini.md [hunk 11]
```diff
@@ -68,25 +68,22 @@ Start with "[ Optimizer ]"
 
 1. ```type = <string>```
 
-   Optimizer type to apply the gradients to weights.
+   Optimizer type to apply the gradients to weights. The default value is adam if the type is not used.
      * adam : Adaptive Moment Estimation
      * sgd : stochastic gradient decent
 
-2. ```learning_rate = <float>```
+2. ```beta1 = <float>```
```
[the remark this answers is not in the document]
Comment (songgot) [reply docs/configuration-ini.md:75]: "Only valid for adam" seems to be what you said.

File: docs/configuration-ini.md [hunk 12]
```diff
@@ -68,25 +68,22 @@ Start with "[ Optimizer ]"
...
+3. ```beta2 = <float>```
 
-   beta2 parameter for adam optimizer. Only valid for adam. 0.999 is default.
+   beta2 parameter for adam optimizer. Only valid for adam. The default value is 0.999.
 
-5. ```epsilon = <float>```
+4. ```epsilon = <float>```
 
-     Epsilon parameter for adam optimizer. Only valid for adam. 1.0e-7 is default.
+   Epsilon parameter for adam optimizer. Only valid for adam. The default value is 1.0e-7 is.
```
[the remark this answers is not in the document]
Comment (songgot) [reply docs/configuration-ini.md:86]: okay

[the remark this answers is not in the document]
Comment (songgot) [reply docs/configuration-ini.md:86]: 1.0e-7  is not the default? okay.

File: docs/configuration-ini.md (hunk 2 above)
Comment (songgot) [reply docs/configuration-ini.md:104]: okay.

File: docs/configuration-ini.md (hunk 8 above)
Comment (songgot) [reply docs/configuration-ini.md:109]: okay

File: docs/configuration-ini.md (hunk 3 above)
Comment (songgot) [reply docs/configuration-ini.md:129]: okay

File: docs/configuration-ini.md (hunk 4 above)
Comment (songgot) [reply docs/configuration-ini.md:161]: okay

File: docs/configuration-ini.md [hunk 13]
```diff
@@ -101,159 +98,395 @@ beta2 = 0.999
...
 
-13. ```stride = <unsigned int>,<unsigned int>```
+   Data path for validation.
 
-     set stride for convolution and pooling layer
+Below is a sample TrainSet section.
 
-14. ```padding = <unsigned int>,<unsigned int>```
+```ini
+# TrainSet Section
```
[the remark this answers is not in the document]
Comment (songgot) [reply docs/configuration-ini.md:164]: okay

File: docs/configuration-ini.md (hunk 5 above)
Comment (songgot) [reply docs/configuration-ini.md:165]: okay

File: docs/configuration-ini.md (hunk 6 above)
[the remark this answers is not in the document]
Comment (songgot) [reply docs/configuration-ini.md:173]: okay

Comment (songgot) [reply docs/configuration-ini.md:173]: okay

File: docs/configuration-ini.md (hunk 7 above)
[the remark this answers is not in the document]
Comment (songgot) [reply docs/configuration-ini.md:183]: okay

File: docs/configuration-ini.md [hunk 14]
```diff
@@ -101,159 +98,395 @@ beta2 = 0.999
...
 
-17. ```flatten = <bool>```
+   Data path for test.
 
-    flattens the output of this layer.
+Below is a sample TrainSet section.
 
-    Enabling this option is equivalent to attaching a flatten layer after the current layer.
+```ini
+# TrainSet Section
```
[the remark this answers is not in the document]
Comment (songgot) [reply docs/configuration-ini.md:186]: okay

File: docs/configuration-ini.md (hunk 9 above)
Comment (songgot) [reply docs/configuration-ini.md:217]: okay

File: docs/configuration-ini.md [hunk 15]
```diff
@@ -344,71 +577,91 @@ activation = softmax
 ```
 
 ### Configuration file example
+Only INI formatted files *.ini is supported to construct a model from a file.
+Special sections [Model], [Optimizers], [train_set], [valid_set], [test_set] are respectively referring to model, optimizer and data provider objects. Rest of INI sections map to a layer. Keys and values from each section set properties of the layer. All keys and values are treated as case-insensitive.
```
[the remark this answers is not in the document]
Comment (songgot) [reply docs/configuration-ini.md:581]: okay

File: docs/configuration-ini.md [hunk 16]
```diff
@@ -344,71 +577,91 @@ activation = softmax
...
-Save_Path = "model.bin"
-batch_size = 32
-beta1 = 0.9
-beta2 = 0.999
-epsilon = 1e-7
+Type = NeuralNetwork # Network Type : Regression, KNN, NeuralNetwork
+Epochs = 1500        # Epochs
+Loss = cross  	     # Loss function : mse (mean squared error)
+                     #                 cross ( for cross entropy )
+#Save_Path = "mnist_model.bin"    # model path to save / read
```
[the remark this answers is not in the document]
Comment (songgot) [reply docs/configuration-ini.md:593]: no, this ini is from ./Application/MNIST.
I will remove this line.

File: docs/configuration-ini.md [hunk 17]
```diff
@@ -344,71 +577,91 @@ activation = softmax
...
-unit = 10
+Type = fully_connected
+input_layers=flatten
+Unit = 10		# Output Layer Dimension ( = Weight Width )
 weight_initializer = xavier_uniform
 bias_initializer = zeros
-activation = softmax
+Activation = softmax 	# activation : sigmoid, softmax
 ```
+The following restrictions must be adhered to:
```
[the remark this answers is not in the document]
Comment (songgot) [reply docs/configuration-ini.md:662]: okay.

File: docs/configuration-ini.md (hunk 10 above)
Comment (songgot) [reply docs/configuration-ini.md:604]: okay

File: docs/configuration-ini.md (hunk 16 above)
[the remark this answers is not in the document]
Comment (songgot) [reply docs/configuration-ini.md:593]: let's remove #

File: docs/configuration-ini.md [hunk 18]
```diff
@@ -344,71 +589,97 @@ activation = softmax
...
 
+# Layer Section : Name
 [conv2d_c1_layer]
-type = conv2d
+Type = conv2d
+input_layers = inputlayer
 kernel_size = 5,5
 bias_initializer=zeros
-activation=sigmoid
+Activation=sigmoid
```
Comment (jijoongmoon) [inline docs/configuration-ini.md:641]: Just for the information, ini file is not type sensitive.

Comment (songgot) [reply docs/configuration-ini.md:641]: okay

**Comment by lhs8928** [comment]
Preload, InputShape, InputLayer, OutputLayer properties in backbone layer is deprecated.
Please delete this properties.(refer #1620)

**Comment by songgot** [comment]
okay

File: docs/configuration-ini.md [hunk 19]
```diff
@@ -53,227 +53,472 @@ Below is sample Network section.
...
 
-   Type of Layer
-     * input : input layer
-     * conv2d : 2D convolution layer
-     * pooling2d : 2D pooling layer
-     * flatten : flatten layer
-     * fully_connected : fully connected layer
-     * batch_normalization : batch normalization layer
-     * activation : activation layer
+2. ```learning_rate = <float>```
```
Comment (lhs8928) [inline docs/configuration-ini.md:111]: 1. learning_rate property can be a vector of float value in step scheduler.
2. Iteration property is missing. This property is used in step scheduler

Comment (songgot) [reply docs/configuration-ini.md:111]: What kind of type for iteration ?

Comment (lhs8928) [reply docs/configuration-ini.md:111]: The type of iteration is unsigend int.
The step scheduler is added recently so there is no unit test for that for now.

Comment (songgot) [reply docs/configuration-ini.md:111]: okay thank you.

File: docs/configuration-ini.md [hunk 20]
```diff
@@ -53,227 +50,484 @@ Below is sample Network section.
...
+   Decay steps. Only valid for exponential.
 
-    ```class Name [i], class name [i+1],...```
+4. ```decay_rate = <float>```
 
-### Layer Section
+   Decay rate
 
-Describe hyper-parameters for layer. Order of layers in the model follows the order of definition of layers here from top to bottom.
+5. ```iteration = <unsigned int>```
```
Comment (lhs8928) [inline docs/configuration-ini.md:128]: iteration could also be a vector of unsigned int

**Review by jijoongmoon — APPROVED** [review]
LGTM

File: docs/configuration-ini.md [hunk 21]
```diff
@@ -53,227 +50,484 @@ Below is sample Network section.
...
+2. ```learning_rate = <float>```
 
-    validation data file path.   The data must be saved as following
+   Initial learning rate to decay.
 
-    ```feature data[i], label data[i], feature data[i+1], label data[i+1], ...```
+   Constant and exponential receive only one float value.
 
-4. ```testdata = <string>```
+   However, Step must receive two or more float values seperated by commas.
```
Comment (wooksong) [inline docs/configuration-ini.md:115]: Typo.

```suggestion
   However, the step must receive two or more float values separated by commas.
```

How about that?

```suggestion
   However, two or more float values separated by commas should be given for the step.
```

File: docs/configuration-ini.md (hunk 20 above)
Comment (songgot) [reply docs/configuration-ini.md:128]: okay

File: docs/configuration-ini.md (hunk 21 above)
Comment (songgot) [reply docs/configuration-ini.md:115]: thank you

**Review by jijoongmoon — APPROVED** [review]
LGTM

**Review by lhs8928 — APPROVED** [review]
LGTM
